source/extensions/filters/http/cache/config.cc

```diff
@@ -2,31 +2,56 @@
 
 #include "source/extensions/filters/http/cache/cache_filter.h"
 
+#include "http_cache.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace Cache {
 
-Http::FilterFactoryCb CacheFilterFactory::createFilterFactoryFromProtoTyped(
-    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
-    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+namespace {
+template <typename FactoryCtx>
+HttpCacheSharedPtr getCache(const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+                            FactoryCtx& ctx) {
   const std::string type{TypeUtil::typeUrlToDescriptorFullName(config.typed_config().type_url())};
-  HttpCacheFactory* const http_cache_factory =
-      Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type);
-  if (http_cache_factory == nullptr) {
+  auto* http_cache_factory = Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type);
+  if (Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type) == nullptr) {
     throw EnvoyException(
         fmt::format("Didn't find a registered implementation for type: '{}'", type));
   }
 
-  auto cache = http_cache_factory->getCache(config, context);
+  auto cache = http_cache_factory->getCache(config, ctx);
+  if (cache == nullptr) {
+    throw EnvoyException(fmt::format("Didn't find HTTP cache for type: '{}', config: '{}'", type,
+                                     config.DebugString()));
+  }
+
+  return cache;
+}
+} // namespace
+
+Http::FilterFactoryCb CacheFilterFactory::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+
+  auto cache = getCache(config, context);
 
-  return [config, stats_prefix, &context,
-          cache](Http::FilterChainFactoryCallbacks& callbacks) -> void {
-    callbacks.addStreamFilter(std::make_shared<CacheFilter>(config, stats_prefix, context.scope(),
-                                                            context.timeSource(), *cache));
+  return [config, stats_prefix, cache,
+          &context](Http::FilterChainFactoryCallbacks& callbacks) -> void {
+    callbacks.addStreamFilter(std::make_shared<CacheFilter>(stats_prefix, context.scope(),
+                                                            context.timeSource(), config, cache));
   };
 }
 
+Router::RouteSpecificFilterConfigConstSharedPtr
+CacheFilterFactory::createRouteSpecificFilterConfigTyped(
+    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+    Server::Configuration::ServerFactoryContext& context, ProtobufMessage::ValidationVisitor&) {
+
+  auto cache = getCache(config, context);
```

Comment:
It should be possible for a cache implementation to have a heavy, global non-route-specific configuration (e.g. RPC server, disk partition, global timeouts), and a lighter, local route-specific configuration (e.g. admission rules, per-route rate limits, default max-age). In other implementations, non-route-specific configuration may just be default values for the route-specific config.

Regardless, the final step of config acceptance has to be per-filter object, because config changes. Regarding the cost of repeatedly parsing cache implementation protobufs, perhaps we need an HttpCacheFactory method that consumes a protobuf and returns an interface object for a parsed form, that we would then pass to the cache implementation?

I definitely don't have the answers figured out, but I'm confident we'll find them.